drivers/sensor/vl53l1x/vl53l1.c

```diff
@@ -0,0 +1,482 @@
+ * short    | 136           | 135
+ * medium   | 290           | 76
+ * long     | 360           | 73
+ */
+static int vl53l1x_set_mode(const struct device *dev,
+							const struct sensor_value *val)
+{
+	struct vl53l1x_data *drv_data = dev->data;
+	VL53L1_Error ret;
+
+	switch (val->val1) {
+	/* short */
+	case 1:
+	/* medium */
+	case 2:
+	/* long */
+	case 3:
+		drv_data->distance_mode = val->val1;
+		break;
+	default:
+		drv_data->distance_mode = VL53L1_DISTANCEMODE_LONG;
+		break;
+	}
+
+	ret = VL53L1_SetDistanceMode(&drv_data->vl53l1x, drv_data->distance_mode);
+	if (ret != VL53L1_ERROR_NONE) {
+		LOG_ERR("[%s] VL53L1_SetDistanceMode return error (%d)", dev->name, ret);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/*
+ * The ROI is a 16x16 grid.
+ * The bottom left is (0,0), top right is (15, 15).
+ * See UM2356 User Manual (VL531X API doc).
+ */
+static int vl53l1x_set_roi(const struct device *dev,
+						   const struct sensor_value *val)
+{
+	struct vl53l1x_data *drv_data = dev->data;
+	VL53L1_Error ret;
+
+	if ((val->val1 < 0) ||
+		(val->val2 < 0) ||
+		(val->val1 > (16 * 16)) ||
+		(val->val2 > (16 * 16)) ||
+		(val->val2 >= val->val1)) {
+		return -EINVAL;
+	}
+
+	/* Map val to pUserROi */
+	VL53L1_UserRoi_t pUserROi = {
+		.TopLeftX = val->val1 % 16,
+		.TopLeftY = (uint8_t)(val->val1 / 16),
+		.BotRightX = val->val2 % 16,
+		.BotRightY = (uint8_t)(val->val2 / 16),
+	};
+
+	ret = VL53L1_SetUserROI(&drv_data->vl53l1x, &pUserROi);
+	if (ret != VL53L1_ERROR_NONE) {
+		LOG_ERR("[%s] VL53L1_SetUserROI return error (%d)", dev->name, ret);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int vl53l1x_get_mode(const struct device *dev,
+							struct sensor_value *val)
+{
+	struct vl53l1x_data *drv_data = dev->data;
+	VL53L1_DistanceModes mode;
+	VL53L1_Error ret;
+
+	ret = VL53L1_GetDistanceMode(&drv_data->vl53l1x, &mode);
+	if (ret != VL53L1_ERROR_NONE) {
+		LOG_ERR("[%s] VL53L1_GetDistanceMode return error (%d)", dev->name, ret);
+		return -EINVAL;
+	}
+
+	/* Mapping is 1:1 with the API */
+	val->val1 = (int32_t)mode;
+	val->val2 = 0;
+	return 0;
+}
+
+static int vl53l1x_get_roi(const struct device *dev,
+						   struct sensor_value *val)
+{
+	struct vl53l1x_data *drv_data = dev->data;
+	VL53L1_Error ret;
+	VL53L1_UserRoi_t pUserROi;
+
+	ret = VL53L1_GetUserROI(&drv_data->vl53l1x, &pUserROi);
+	if (ret != VL53L1_ERROR_NONE) {
+		LOG_ERR("[%s] VL53L1_GetUserROI return error (%d)", dev->name, ret);
+		return -EINVAL;
+	}
+
+	/* Map pUserROi to val */
+	val->val1 = (int32_t)((16 * pUserROi.TopLeftY) + pUserROi.TopLeftX);
+	val->val2 = (int32_t)((16 * pUserROi.BotRightY) + pUserROi.BotRightX);
+	return 0;
+}
+
+static int vl53l1x_sample_fetch(const struct device *dev,
+								enum sensor_channel chan)
+{
+	struct vl53l1x_data *drv_data = dev->data;
+	VL53L1_Error ret;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_DISTANCE);
+
+	if (!drv_data->initialized) {
+		int init_ret = vl53l1x_initialize(dev);
```

Comment:
Can this happen?  It looks like the driver init does the initialization already.